coordiante_transform/src/coordinate_transform_component.cpp

```diff
@@ -0,0 +1,187 @@
+
+  void DetectionCallback(const stsl_interfaces::msg::TagArray::SharedPtr tag_array_msg)
+  {
+    std::string tf_error_string;
+    if (!tf_buffer_.canTransform(
+            "base_footprint", tag_array_msg->header.frame_id,
+            tag_array_msg->header.stamp, tf2::durationFromSec(0.1),
+            &tf_error_string))
+    {
+      RCLCPP_WARN_THROTTLE(
+              get_logger(), *get_clock(), 1000, "Could not lookup transform. %s",
+              tf_error_string);
+      return;
+    }
+
+    // find the transform from the camera to base footprint
+    const auto tf_transform =
+            tf_buffer_.lookupTransform("base_footprint", "camera_link", tag_array_msg->header.stamp);
+    const Eigen::Isometry3d eigen_transform = tf2::transformToEigen(tf_transform);
+
+    // creates a matrix that goes from camera to standard ROS coordinates
+    Eigen::Matrix4d optical = getRotMatForOpticalFrame();
+
+    // BEGIN STUDENT CODE
+    // create a new tag array message
+    stsl_interfaces::msg::TagArray new_tag_array_msg;
+    // copy the header information
+    new_tag_array_msg.header = tag_array_msg->header;
+    // change the frame_id to be the correct reference frame
+    new_tag_array_msg.header.frame_id = "base_footprint";
+    // END STUDENT CODE
+
+    // BEGIN STUDENT CODE
+    // iterate over each tag to and transform it into body frame
+    for(int i = 0; i < tag_array_msg->tags.size(); i++) {
+      stsl_interfaces::msg::Tag new_tag;
+      new_tag.id = tag_array_msg->tags[i].id;
+
+      geometry_msgs::msg::Point old_tag_position = tag_array_msg->tags[i].pose.position;
+
+      // extract the important parts of the pose
+      Eigen::Vector4d pose = Eigen::Vector4d(old_tag_position.x, old_tag_position.y, old_tag_position.z, 1);
+
+      // apply the transform
+      pose = eigen_transform.matrix() * optical * pose;
+
+      new_tag.pose.position.x = pose(0);
+      new_tag.pose.position.y = pose(1);
+      new_tag.pose.position.z = pose(2);
+
+      // get the orientation of the tag
+      Eigen::Matrix4d tag_orientation = convertQuatToRotMat(tag_array_msg->tags[i].pose.orientation);
+      tag_orientation = eigen_transform.matrix() * optical * tag_orientation;
+      new_tag.pose.orientation = convertRotMatToQuat(tag_orientation);
+
+      new_tag_array_msg.tags.push_back(new_tag);
+    }
+    tag_publisher_->publish(new_tag_array_msg);
+    // END STUDENT CODE
+
+    // visualization code
+    visualization_msgs::msg::MarkerArray marker_array;
+    for(stsl_interfaces::msg::Tag tag : new_tag_array_msg.tags)
+    {
+      visualization_msgs::msg::Marker marker;
+      marker.header.frame_id = "base_footprint";
+      marker.header.stamp = tag_array_msg->header.stamp;
+      marker.id = tag.id;
+      marker.type = visualization_msgs::msg::Marker::CUBE;
+      marker.action = visualization_msgs::msg::Marker::ADD;
+      marker.lifetime = rclcpp::Duration(0.25s);
```

Comment:
For discussion: Should we just use the seconds+nanoseconds constructor instead and avoid having to import the literals namespace above?
```suggestion
      marker.lifetime = rclcpp::Duration(0, 250000000);
```

Reply:
I think for week one yes, but can re evaluate later on.